org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java

```diff
@@ -637,30 +643,106 @@ private void collectionRegionProposals(ICompletionRequest request, ICompletionRe
-				endProposal.setDocumentation("Folding Region End");
+				endProposal.setDocumentation("Insert Folding Region End");
 				endProposal.setFilterText(match.group());
 				endProposal.setSortText("zb");
+				endProposal.setKind(CompletionItemKind.Snippet);
+				endProposal.setInsertTextFormat(InsertTextFormat.PlainText);
 				response.addCompletionAttribute(endProposal);
 			}
 		} catch (BadLocationException e) {
 			LOGGER.log(Level.SEVERE, "While performing collectRegionCompletion", e);
 		}
 	}
 
+	private void collectCDATACompletion(ICompletionRequest request, ICompletionResponse response) {
+		try {
+			boolean isSnippetsSupported = request.isCompletionSnippetsSupported();
+			InsertTextFormat insertFormat = request.getInsertTextFormat();
+
+			DOMDocument document = request.getXMLDocument();
+
+			String filter = "<!CDATA";
+			int startOffset = StringUtils.findExprBeforeAt(document.getText(), filter, request.getOffset());
+			if (startOffset == -1) {
+				startOffset = StringUtils.findExprBeforeAt(document.getText(), "CDATA", request.getOffset());
+				if (startOffset != - -1) {
+					filter = "CDATA";
+				}
+			}
+			if (startOffset == -1) {
+				startOffset = request.getOffset();
+			}
+			int endOffset = request.getOffset();
+			Range editRange = getReplaceRange(startOffset + 1, endOffset, request);
+
+			// Insert CDATA Section
+			CompletionItem cdataProposal = new CompletionItem("cdata");
+			cdataProposal.setKind(CompletionItemKind.Snippet);
+			cdataProposal.setFilterText(filter);
+			cdataProposal.setSortText("zc");
+			cdataProposal.setInsertTextFormat(insertFormat);
+
+			String textEdit = isSnippetsSupported ? "<![CDATA[ $1]]>" : "<![CDATA[ ]]>";
+			cdataProposal.setTextEdit(new TextEdit(editRange, textEdit));
+			cdataProposal.setDocumentation("Insert <![CDATA[ ]]>");
+			response.addCompletionItem(cdataProposal);
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "While performing collectCDATACompletion", e);
+		}
+	}
+
+	private void collectCommentCompletion(ICompletionRequest request, ICompletionResponse response) {
+		try {
+			boolean isSnippetsSupported = request.isCompletionSnippetsSupported();
+			InsertTextFormat insertFormat = request.getInsertTextFormat();
+
+			DOMDocument document = request.getXMLDocument();
+			String filter = "<!--";
+			int startOffset = StringUtils.findExprBeforeAt(document.getText(), filter, request.getOffset());
+			if (startOffset == -1) {
+				startOffset = StringUtils.findExprBeforeAt(document.getText(), "comment", request.getOffset());
+				if (startOffset != - -1) {
+					filter = "comment";
+				}
+			}
+			if (startOffset == -1) {
+				startOffset = request.getOffset();
+			}
+			int endOffset = request.getOffset();
+			Range editRange = getReplaceRange(startOffset + 1, endOffset, request);
+
+			// Insert Comment Section
+			CompletionItem cdataProposal = new CompletionItem("comment");
+			cdataProposal.setKind(CompletionItemKind.Snippet);

```

Comment:
isSnippetsSupported?

Reply:
It's just for icon. Snippet support is for insert text format